File: middleware/tests/admin/test_onboard.py
# The MIT License (MIT)
#
# Copyright (c) 2021 RSK Labs Ltd
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from types import SimpleNamespace
from unittest import TestCase
from unittest.mock import Mock, call, patch, mock_open
from admin.certificate import HSMCertificate, HSMCertificateElement
from admin.misc import AdminError
from admin.onboard import do_onboard
import json
from ledger.hsm2dongle import HSM2Dongle, HSM2DongleError

import logging


logging.disable(logging.CRITICAL)


@patch("sys.stdout.write")
@patch("time.sleep")
@patch("admin.onboard.info")
@patch("admin.onboard.get_hsm")
class TestOnboard(TestCase):
    VALID_PIN = "1234ABCD"
    INVALID_PIN = "123456789"

    DEVICE_KEY = {
        "pubkey": "this-is-the-public-key",
        "message": "1122334455",
        "signature": "aabbccddee",
    }

    ENDORSEMENT_KEY = {
        "pubkey": "this-is-another-public-key",
        "message": "5544332211",
        "signature": "eeddccbbaa",
    }

    INVALID_KEY = {
        "pubkey": "this-is-the-public-key",
        "message": "invalid-message",
        "signature": "invalid-signature",
    }

    def setUp(self):
        self.certificate_path = "cert-path"
        options = {
            "pin": self.VALID_PIN,
            "output_file_path": self.certificate_path,
            "any_pin": False,
            "no_exec": False,
            "verbose": False
        }
        self.default_options = SimpleNamespace(**options)

        self.expected_cert = HSMCertificate()
        self.expected_cert.add_element(
            HSMCertificateElement({
                "name": "attestation",
                "message": self.ENDORSEMENT_KEY["message"],
                "signature": self.ENDORSEMENT_KEY["signature"],
                "signed_by": "device",
            })
        )
        self.expected_cert.add_element(
            HSMCertificateElement({
                "name": "device",
                "message": self.DEVICE_KEY["message"],
                "signature": self.DEVICE_KEY["signature"],
                "signed_by": "root",
            })
        )
        self.expected_cert.add_target("attestation")
        self.dongle = Mock()

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard(self, readline, get_hsm_unlock, get_admin_hsm,
                     get_hsm_onboard, info_mock, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock(return_value=self.DEVICE_KEY)
        self.dongle.setup_endorsement_key = Mock(return_value=self.ENDORSEMENT_KEY)
        self.dongle.handshake = Mock()
        self.dongle.onboard = Mock()
        readline.return_value = "yes\n"

        with patch("builtins.open", mock_open()) as file_mock:
            do_onboard(self.default_options)

        self.assertEqual(info_mock.call_args_list[5][0][0], "Onboarded: No")
        self.assertEqual(info_mock.call_args_list[10][0][0], "Onboarded")
        self.assertEqual(info_mock.call_args_list[14][0][0], "Device key gathered")
        self.assertEqual(info_mock.call_args_list[16][0][0],
                         "Attestation key setup complete")

        self.assertEqual([call(self.certificate_path, "w")], file_mock.call_args_list)
        self.assertEqual([call("%s\n" %
                               json.dumps(self.expected_cert.to_dict(), indent=2))],
                         file_mock.return_value.write.call_args_list)
        self.assertTrue(self.dongle.onboard.called)
        self.assertTrue(self.dongle.handshake.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_already_onboarded(self, readline, get_hsm_unlock, get_admin_hsm,
                                       get_hsm_onboard, info_mock, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(return_value=True)

        with self.assertRaises(AdminError) as e:
            do_onboard(self.default_options)

        self.assertEqual(info_mock.call_args_list[5][0][0], "Onboarded: Yes")
        self.assertEqual(e.exception.args[0], "Device already onboarded")
        self.assertFalse(self.dongle.onboard.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_onboard_error(self, readline, get_hsm_unlock, get_admin_hsm,
                                   get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(return_value=False)
        self.dongle.get_device_key = Mock()
        self.dongle.setup_endorsement_key = Mock()
        self.dongle.handshake = Mock()
        self.dongle.onboard = Mock(side_effect=HSM2DongleError("error-msg"))
        readline.return_value = "yes\n"

        with patch("builtins.open", mock_open()) as file_mock:
            with self.assertRaises(HSM2DongleError) as e:
                do_onboard(self.default_options)

        self.assertTrue(self.dongle.onboard.called)
        self.assertEqual("error-msg", str(e.exception))
        self.assertFalse(self.dongle.get_device_key.called)
        self.assertFalse(self.dongle.setup_endorsement_key.called)
        self.assertFalse(self.dongle.handshake.called)
        self.assertFalse(file_mock.return_value.write.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_handshake_error(self, readline, get_hsm_unlock, get_admin_hsm,
                                     get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock()
        self.dongle.setup_endorsement_key = Mock()
        self.dongle.handshake = Mock(side_effect=HSM2DongleError("error-msg"))
        self.dongle.onboard = Mock()
        readline.return_value = "yes\n"

        with patch("builtins.open", mock_open()) as file_mock:
            with self.assertRaises(HSM2DongleError) as e:
                do_onboard(self.default_options)

        self.assertEqual("error-msg", str(e.exception))
        self.assertTrue(self.dongle.onboard.called)
        self.assertTrue(self.dongle.handshake.called)
        self.assertFalse(self.dongle.get_device_key.called)
        self.assertFalse(self.dongle.setup_endorsement_key.called)
        self.assertFalse(file_mock.return_value.write.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_getkey_error(self, readline, get_hsm_unlock, get_admin_hsm,
                                  get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock(side_effect=HSM2DongleError("error-msg"))
        self.dongle.setup_endorsement_key = Mock()
        self.dongle.handshake = Mock()
        self.dongle.onboard = Mock()
        readline.return_value = "yes\n"

        with patch("builtins.open", mock_open()) as file_mock:
            with self.assertRaises(HSM2DongleError) as e:
                do_onboard(self.default_options)

        self.assertEqual("error-msg", str(e.exception))
        self.assertTrue(self.dongle.onboard.called)
        self.assertTrue(self.dongle.handshake.called)
        self.assertTrue(self.dongle.get_device_key.called)
        self.assertFalse(self.dongle.setup_endorsement_key.called)
        self.assertFalse(file_mock.return_value.write.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_setupkey_error(self, readline, get_hsm_unlock, get_admin_hsm,
                                    get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock()
        self.dongle.setup_endorsement_key = Mock(side_effect=HSM2DongleError("error-msg"))
        self.dongle.handshake = Mock()
        self.dongle.onboard = Mock()
        readline.return_value = "yes\n"

        with patch("builtins.open", mock_open()) as file_mock:
            with self.assertRaises(HSM2DongleError) as e:
                do_onboard(self.default_options)

        self.assertEqual("error-msg", str(e.exception))
        self.assertTrue(self.dongle.onboard.called)
        self.assertTrue(self.dongle.handshake.called)
        self.assertTrue(self.dongle.get_device_key.called)
        self.assertTrue(self.dongle.setup_endorsement_key.called)
        self.assertFalse(file_mock.return_value.write.called)

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_user_cancelled(self, readline, hsm_unlock, hsm_admin,
                                    hsm_onboard, *_):
        hsm_onboard.return_value = self.dongle
        hsm_unlock.return_value = self.dongle
        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(return_value=False)
        self.dongle.onboard = Mock()
        hsm_admin.return_value = self.dongle
        readline.return_value = "no\n"

        with patch("builtins.open", mock_open()) as file_mock:
            with self.assertRaises(AdminError) as e:
                do_onboard(self.default_options)

        self.assertEqual("Cancelled by user", str(e.exception))
        self.assertFalse(self.dongle.onboard.called)
        self.assertFalse(self.dongle.get_device_key.called)
        self.assertFalse(self.dongle.setup_endorsement_key.called)
        self.assertFalse(file_mock.return_value.write.called)

    @patch("sys.stdin.readline")
    def test_onboard_no_output_file(self, readline, get_hsm, *_):
        readline.return_value = "yes\n"
        get_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(return_value=False)
        self.dongle.onboard = Mock()

        options = self.default_options
        options.output_file_path = None
        with self.assertRaises(AdminError) as e:
            do_onboard(options)

        self.assertEqual("No output file path given", str(e.exception))
        self.assertFalse(self.dongle.onboard.called)

    def test_onboard_invalid_pin(self, *_):
        options = self.default_options
        options.pin = self.INVALID_PIN
        with self.assertRaises(AdminError) as e:
            do_onboard(options)

        self.assertTrue(str(e.exception).startswith("Invalid pin given."))

    def test_onboard_invalid_mode(self, get_hsm, *_):
        get_hsm.return_value = self.dongle
        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.SIGNER)
        self.dongle.is_onboarded = Mock(return_value=False)

        with self.assertRaises(AdminError) as e:
            do_onboard(self.default_options)

        self.assertTrue(str(e.exception).startswith("Device not in bootloader mode."))

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_invalid_device_key(self, readline, get_hsm_unlock, get_admin_hsm,
                                        get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock(return_value=self.INVALID_KEY)
        self.dongle.setup_endorsement_key = Mock(return_value=self.ENDORSEMENT_KEY)
        readline.return_value = "yes\n"

        with self.assertRaises(ValueError) as e:
            do_onboard(self.default_options)

        self.assertEqual(("Missing or invalid message for HSM certificate element"
                          " device"), str(e.exception))

    @patch("admin.onboard.get_admin_hsm")
    @patch("admin.unlock.get_hsm")
    @patch("sys.stdin.readline")
    def test_onboard_invalid_attestation_key(self, readline, get_hsm_unlock,
                                             get_admin_hsm, get_hsm_onboard, *_):
        get_hsm_onboard.return_value = self.dongle
        get_hsm_unlock.return_value = self.dongle
        get_admin_hsm.return_value = self.dongle

        self.dongle.get_current_mode = Mock(return_value=HSM2Dongle.MODE.BOOTLOADER)
        self.dongle.is_onboarded = Mock(side_effect=[False, True])
        self.dongle.get_device_key = Mock(return_value=self.DEVICE_KEY)
        self.dongle.setup_endorsement_key = Mock(return_value=self.INVALID_KEY)
        readline.return_value = "yes\n"

        with self.assertRaises(ValueError) as e:
            do_onboard(self.default_options)

        self.assertEqual(("Missing or invalid message for HSM certificate element"
                          " attestation"), str(e.exception))

File: ledger/src/ui/test/onboard/test_onboard.c
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 RSK Labs Ltd
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <stdbool.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>

#include "cx.h"
#include "defs.h"
#include "ui_err.h"
#include "onboard.h"
#include "os.h"
#include "pin.h"
#include "assert_utils.h"
#include "apdu_utils.h"
#include "ui_instructions.h"

/**
 * Mock variables used to assert function calls
 */
static unsigned char G_mock_seed[32];
static unsigned char G_pin_buffer[10];
static unsigned char G_device_pin[10];
static unsigned char G_global_seed[257];
static bool G_is_pin_valid;
static bool G_device_unlocked;
static bool G_device_onboarded;
static bool G_wiped_while_locked;

static void setup() {
    memset(G_mock_seed, 0, sizeof(G_mock_seed));
    memset(G_pin_buffer, 0, sizeof(G_pin_buffer));
    memset(G_device_pin, 0, sizeof(G_device_pin));
    G_is_pin_valid = false;
    G_device_unlocked = false;
    G_device_onboarded = false;
    G_wiped_while_locked = false;
}

void mock_cx_rng(const unsigned char *data, unsigned int len) {
    for (unsigned int i = 0; i < len; i++) {
        G_mock_seed[i] = data[i];
    }
}

unsigned char *cx_rng(unsigned char *buffer, unsigned int len) {
    // Mock 32 random bytes
    for (int i = 0; i < len; i++) {
        buffer[i] = G_mock_seed[i];
    }
    return 0;
}

unsigned int bolos_ux_mnemonic_from_data(unsigned char *in,
                                         unsigned int inLength,
                                         unsigned char *out,
                                         unsigned int outLength) {
    const char mnemonics_prefix[] = "mnemonics-generated-from-";
    int len = sizeof(mnemonics_prefix) - 1;
    assert(outLength >= len + inLength);
    memcpy(out, mnemonics_prefix, len);
    memcpy(out + len, in, inLength);
    return len + inLength;
}

/**
 * Mock calls to other modules
 */
bool is_pin_valid() {
    return G_is_pin_valid;
}

void set_device_pin() {
    // NOTE: set_device_pin skips the prepended length
    memcpy(G_device_pin, G_pin_buffer + 1, sizeof(G_pin_buffer) - 1);
}

unsigned int unlock_with_pin(bool prepended_length) {
    const char *pin_buffer;
    if (prepended_length) {
        pin_buffer = (const char *)(G_pin_buffer + 1);
    } else {
        pin_buffer = (const char *)G_pin_buffer;
    }

    G_device_unlocked = (0 == strcmp((const char *)G_device_pin, pin_buffer));
    return G_device_unlocked;
}

unsigned int update_pin_buffer(volatile unsigned int rx) {
    unsigned char index = APDU_AT(2);
    unsigned char val = APDU_AT(3);
    G_pin_buffer[index] = val;
    G_pin_buffer[index + 1] = 0;
    return 3;
}

/**
 * Mock OS calls
 */
void os_global_pin_invalidate(void) {
    G_device_unlocked = false;
}

void os_perso_derive_and_set_seed(unsigned char identity,
                                  const char *prefix,
                                  unsigned int prefix_length,
                                  const char *passphrase,
                                  unsigned int passphrase_length,
                                  const char *words,
                                  unsigned int words_length) {
    sprintf((char *)G_global_seed, "seed-generated-from-%s", words);
}

void os_perso_finalize(void) {
    G_device_onboarded = true;
}

unsigned int os_perso_isonboarded(void) {
    return G_device_onboarded;
}

void os_perso_wipe() {
    if (!G_device_unlocked) {
        G_wiped_while_locked = true;
    }
    // wipe global pin, seed and state
    memset(G_device_pin, 0x0, sizeof(G_device_pin));
    memset(G_global_seed, 0x0, sizeof(G_global_seed));
    G_device_unlocked = false;
    G_device_onboarded = false;
}

// Helpers for RSK commands
static void send_rsk_pin_cmd(const char *pin) {
    unsigned int rx = 4;
    for (int i = 0; i < strlen(pin); i++) {
        SET_APDU_AT(0, CLA);
        SET_APDU_AT(1, RSK_PIN_CMD);
        SET_APDU_AT(2, i);
        SET_APDU_AT(3, pin[i]);
        assert(3 == update_pin_buffer(rx));
    }
}

static void send_rsk_seed_cmd(onboard_t *ctx, const unsigned char *host_seed) {
    unsigned int rx = 4;
    for (int i = 0; i < SEED_LENGTH; i++) {
        SET_APDU_AT(0, CLA);
        SET_APDU_AT(1, RSK_SEED_CMD);
        SET_APDU_AT(2, i);
        SET_APDU_AT(3, host_seed[i]);
        assert(0 == set_host_seed(rx, ctx));
    }
}

char words_buffer[] = "words_buffer";
char seed[] = "seed_buffer";

void test_reset_onboard_ctx() {
    printf("Test reset onboard context...\n");

    setup();
    onboard_t onboard_ctx;
    memcpy(onboard_ctx.words_buffer, words_buffer, sizeof(words_buffer));
    memcpy(onboard_ctx.seed, seed, sizeof(seed));

    reset_onboard_ctx(&onboard_ctx);

    ASSERT_ARRAY_CLEARED(onboard_ctx.words_buffer);
    ASSERT_ARRAY_CLEARED(onboard_ctx.seed);
    assert(0 == onboard_ctx.words_buffer_length);
}

void test_set_host_seed() {
    printf("Test set host seed...\n");

    setup();
    onboard_t onboard_ctx;
    reset_onboard_ctx(&onboard_ctx);
    // mock 32 bytes random host seed
    const char host_seed[] = {0x9c, 0xc9, 0x8b, 0xde, 0x40, 0xb3, 0x33, 0xbc,
                              0x34, 0xf2, 0xae, 0x6f, 0xe6, 0x09, 0x1f, 0x57,
                              0xfd, 0x4b, 0x6d, 0xcd, 0xc0, 0x62, 0x61, 0x53,
                              0x03, 0xf0, 0xef, 0x03, 0xce, 0x84, 0x15, 0xd3};
    unsigned int rx = 4;
    for (int i = 0; i < strlen(host_seed); i++) {
        SET_APDU_AT(2, i);
        SET_APDU_AT(3, host_seed[i]);
        assert(0 == set_host_seed(rx, &onboard_ctx));
    }
    ASSERT_MEMCMP(host_seed, onboard_ctx.host_seed, SEED_LENGTH);
}

void test_onboard_device() {
    printf("Test onboard device...\n");

    setup();
    onboard_t onboard_ctx;
    reset_onboard_ctx(&onboard_ctx);
    // mock 32 bytes random host seed
    const unsigned char host_seed[] = {
        0x9c, 0xc9, 0x8b, 0xde, 0x40, 0xb3, 0x33, 0xbc, 0x34, 0xf2, 0xae,
        0x6f, 0xe6, 0x09, 0x1f, 0x57, 0xfd, 0x4b, 0x6d, 0xcd, 0xc0, 0x62,
        0x61, 0x53, 0x03, 0xf0, 0xef, 0x03, 0xce, 0x84, 0x15, 0xd3};
    // mock 32 bytes handom seed
    const unsigned char seed[] = {
        0x21, 0x9f, 0x24, 0x3a, 0xf4, 0x23, 0x2e, 0x26, 0x0b, 0x37, 0xfe,
        0xf4, 0xd8, 0xc8, 0xf4, 0x88, 0xa5, 0x3a, 0x36, 0xd7, 0xa8, 0xa2,
        0xd5, 0x42, 0x8f, 0x57, 0xb8, 0x92, 0x79, 0x7f, 0xb0, 0xd3};

    // Set device pin (with prepended length)
    send_rsk_pin_cmd("X1234567a");
    // Set host seed
    send_rsk_seed_cmd(&onboard_ctx, host_seed);

    G_is_pin_valid = true;
    mock_cx_rng(seed, SEED_LENGTH);
    assert(3 == onboard_device(&onboard_ctx));
    ASSERT_APDU("\x80\x02\x01");
    assert(G_device_unlocked);
    assert(G_device_onboarded);
    assert(G_wiped_while_locked);

    ASSERT_STR_EQUALS("1234567a", G_device_pin);
    // "seed-generated-from-" + "mnemonics-generated-from-" + host_seed XOR seed
    ASSERT_MEMCMP(
        "seed-generated-from-mnemonics-generated-from-"
        "\xbd\x56\xaf\xe4\xb4\x90\x1d\x9a\x3f\xc5\x50\x9b\x3e\xc1\xeb\xdf\x58"
        "\x71\x5b\x1a\x68\xc0\xb4\x11\x8c\xa7\x57\x91\xb7\xfb\xa5\x00",
        G_global_seed,
        77);

    // Make sure all mnemonic and seed information is wiped after onboard_device
    ASSERT_ARRAY_CLEARED(onboard_ctx.words_buffer);
    ASSERT_ARRAY_CLEARED(onboard_ctx.seed);
    assert(0 == onboard_ctx.words_buffer_length);
}

void test_onboard_device_invalid_pin() {
    printf("Test onboard device (invalid pin)...\n");

    setup();
    onboard_t onboard_ctx;
    reset_onboard_ctx(&onboard_ctx);
    // mock 32 bytes random host seed
    const unsigned char host_seed[] = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

    // wrong device pin (without prepended length)
    send_rsk_pin_cmd("1234567a");
    // Set host seed
    send_rsk_seed_cmd(&onboard_ctx, host_seed);

    BEGIN_TRY {
        TRY {
            G_is_pin_valid = false;
            onboard_device(&onboard_ctx);
            // onboard_device should throw ERR_UI_INVALID_PIN
            ASSERT_FAIL();
        }
        CATCH(ERR_UI_INVALID_PIN) {

            return;
            assert(!G_device_onboarded);
            assert(!G_device_unlocked);
            ASSERT_ARRAY_CLEARED(G_device_pin);
            ASSERT_ARRAY_CLEARED(G_global_seed);
        }
        CATCH_OTHER(e) {
            ASSERT_FAIL();
        }
        FINALLY {
        }
    }
    END_TRY;
}

void test_is_onboarded() {
    printf("Test is onboarded...\n");

    setup();
    onboard_t onboard_ctx;
    reset_onboard_ctx(&onboard_ctx);
    // mock 32 bytes random host seed
    const unsigned char host_seed[] = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

    // Set device pin (with prepended length)
    send_rsk_pin_cmd("X1234567a");
    // Set host seed
    send_rsk_seed_cmd(&onboard_ctx, host_seed);

    G_device_onboarded = true;
    assert(5 == is_onboarded());
    ASSERT_APDU("\x80\x01\x04\x00\x00");

    G_device_onboarded = false;
    assert(5 == is_onboarded());
    ASSERT_APDU("\x80\x00\x04\x00\x00");
}

int main() {
    test_reset_onboard_ctx();
    test_set_host_seed();
    test_onboard_device();
    test_onboard_device_invalid_pin();
    test_is_onboarded();

    return 0;
}
